### ISSUE TITLE
Solve interpolation still wrong

### ISSUE THREAD

**Issue by thofma** [Nemocas/Nemo.jl#351]
In the following example, interpolation does not encounter an impossible inverse, but the solution is still incorrect:
```julia
julia> S, x = PolynomialRing(ResidueRing(JuliaZZ, 211*17), "x")
(Univariate Polynomial Ring in x over Residue ring of Integers modulo 3587, x)

julia> M = matrix(S, 3, 3, [3446*x+102, 22*x^2+3435*x+3483, 3425*x^3+3430*x^2+74*x+33, 3070*x+374, 132*x^2+2983*x+3187, 2888*x^3+2934*x^2+169*x+112, 3258*x+238, 110*x^2+3179*x+3323, 3089*x^3+3127*x^2+60*x+67])

julia> b = matrix(S, 3, 2, [3568*x^3+3564*x^2+3582*x+26, 3552, 3561*x+15, 40, 3561*x+3575, 3548*x^3+25*x^2+11*x+25])

julia> y, d = Nemo.Generic.solve_interpolation(M, b)

julia> M * y == d * b
false
```
All the solve functions are pretty unreliable (wrong), when it comes to non-integral domains.

**Comment by wbhart** [comment]
This appears to now raise an exception due to a Singular matrix. The code checks whether it is singular for a certain number of values (the bound for which is computed), then if it is for each of them, it raises and exception and says the matrix is singular.

I would guess this is because the bound is violated over a non-integral domain, which is interesting. I guess it could be Singular modulo one of the factors of the moduli. It's quite surprising that the code detects this though.

**Comment by wbhart** [comment]
Yes, so if we work mod 35, the following matrix will cause the code to say the matrix is singular. But in fact, 4, 9, 14, 19, 24, 29 and 34 are all evaluation points that make the matrix non-singular mod both of the divisors of 35. Therefore I think solve_interpolation should succeed in this case, but does not.

[32*x^2+16*x+20 17*x^2+25*x+30]
[34*x^2+29*x+21 5*x^2+3*x+13]

As to how we would even think about fixing this, I don't know. I think the algorithm might just be invalid for non-integral domains. 

For Z/nZ, perhaps it is possible to multiply the bound by some factor based on the largest possible number of factors of the modulus and have it keep trying to ensure that it really is Singular. But this might be a very large bound.

**Comment by wbhart** [comment]
Note the determinant is not zero mod 5 or 7 in this example.

**Comment by wbhart** [comment]
I found an improvement which works in many more cases than before:

* I changed the error message in solve_interpolate to "Impossible inverse or too many singular matrices in solve_interpolate"
* If *any* of the evaluation points succeeds, we do not raise an exception, as the original matrix is not singular

This causes the function to succeed over Z/nZ for vastly more cases than before.

Of course, technically, we should only be raising an exception in solve_interpolate if all the points yield singular matrices. And technically, even if "bound" evaluated matrices yield singular matrices, this doesn't imply the original matrix is singular, as it might just be singular mod different primes.

But at least the error message will be correct, and the function will fail much less often now.

**Comment by wbhart** [comment]
Actually, both solve_ff and solve_interpolate fail for non-integral domains.

Here is a small example that fails for solve_interpolate:

```
S, x = PolynomialRing(ResidueRing(ZZ, 3*5), "x")
R = MatrixSpace(S, 2, 2)
U = MatrixSpace(S, 2, 1)

M = R([4*x+3, 10*x+10, 13*x+6, 12*x+9])
b = U([6*x+10, 3*x+13])

x, d = Generic.solve_interpolation(M, b)

M*x == d*b
```

The value of x returned here is:

```
[2*x^2+14*x]
[9*x^2+9]
```

The correct answer is:

```
[12*x^2+14*x+5]
[9*x^2+9]
```

So, either we have a bug somewhere, or the algorithm simply doesn't work for non-integral domains.

**Comment by wbhart** [comment]
I think the problem is the LU decomposition can give meaningless output.

It's supposed to give L and U and a permutation p such that p(M) = L*D*U where the matrix D is diagonal with entries that are constructed from inverses of diagonal elements of U. The only problem is, those may not be invertible, (which means some data has been lost).

Bizarrely, for the original problem, just inverting the matrix M and taking r = inv(M)*b seems to work just fine.
